qiskit_addon_utils/exp_vals/expectation_values.py

```diff
@@ -0,0 +1,255 @@
+        ddof = 1 
+        ## FIXME
+        variances = np.nan #(1 - means**2) / (num_kept[:,None]-ddof)
+
+        ## AVERAGE OVER SPECIFIED AXES ("TWIRLS"):
+        # Update indexing since we already sliced away meas_basis axis:
+        avg_axis_ = tuple(a if a < meas_basis_idx else a-1 for a in avg_axis)
+        
+        # Will weight each twirl by its fraction of kept shots.
+        # If no postselection, weighting reduces to dividing by num_twirls:
+        weights = num_kept[:, None] / np.sum(num_kept, axis=avg_axis_)
+        num_minus = np.count_nonzero(signs, axis=avg_axis_)
+        num_plus = np.count_nonzero(~signs, axis=avg_axis_)
+        num_twirls = num_plus + num_minus
+        gamma_approx = num_twirls / (num_plus - num_minus)
+        means = gamma_approx * np.sum(means * weights, axis=avg_axis_)
+        # Propagate uncertainties:
+        variances = gamma_approx**2 * np.sum(variances * weights**2, axis=avg_axis_)
+        mean_each_observable += means
+        var_each_observable += variances
+
+    mean_and_var_each_observable = list(zip(mean_each_observable.tolist(), var_each_observable.tolist()))
+
+    return mean_and_var_each_observable
+
+
+def apply_postselect_mask(
+    bool_array: np.ndarray[np._bool],
+    basis_dict: dict[Pauli, list[SparseObservable]],
+    postselect_mask: np.ndarray[np._bool],
+):
+    """Applies postselection mask in preparation for computing expectation values.
+
+    Args:
+        bool_array: Boolean array, presumably representing data from measured qubits.
+            The last two axes are the number of shots and number of classical bits, respectively.
+        basis_dict: This dict encodes how the data in `bool_array` should be used to estimate the desired list of Pauli observables.
+            Similar to `basis_dict` arg of `expectation_values()`, but here Pauli components and coefficients must be represented as
+            `SparseObservable` instead of `SparsePauliOp`, and `None` may not be used as a placeholder for the zero operator.
+        postselect_mask: Boolean array used for postselection. `True` (`False`) indicates a shot accepted (rejected) by postselection.
+            Shape must be `bool_array.shape[:-1]`.
+
+    Returns:
+        - A copy of `bool_array` with an extra classical bit appended indicating the postselection mask.
+        - A copy of `basis_dict` where each observable term has had an extra `1` appended.
+        - An array tabulating how many shots were kept for each circuit configuration. When computing expectation values,
+            this must be used to correct for the number of shots included in each average.
+    """
+    if postselect_mask is not None:
+        # Projector will ignore shots where ps bit is 0 when computing expectation
+        # (though we will need to correct the shot counts later on)
+        basis_dict = {
+            basis: [obs.expand("1") for obs in diag_obs_list]
```

Comment:
I like the use of the `"1"` projector from the `SparseObservable` alphabet (also `Z` in `apply_pec_signs` below. But this was not immediately obvious to me, so adding a comment to explicitly call out what `obs.expand("1")` is actually doing would be useful (same below in `apply_pec_signs`).